examples/fluids/qfunctions/stg_shur14.h

```diff
@@ -0,0 +1,133 @@
+// Copyright (c) 2017, Lawrence Livermore National Security, LLC. Produced at
+// the Lawrence Livermore National Laboratory. LLNL-CODE-734707. All Rights
+// reserved. See files LICENSE and NOTICE for details.
+//
+// This file is part of CEED, a collection of benchmarks, miniapps, software
+// libraries and APIs for efficient high-order finite element and spectral
+// element discretizations for exascale applications. For more information and
+// source code availability see http://github.com/ceed.
+//
+// The CEED research is supported by the Exascale Computing Project 17-SC-20-SC,
+// a collaborative effort of two U.S. Department of Energy organizations (Office
+// of Science and the National Nuclear Security Administration) responsible for
+// the planning and preparation of a capable exascale ecosystem, including
+// software, applications, hardware, advanced system engineering and early
+// testbed platforms, in support of the nation's exascale computing imperative.
+
+/// @file
+/// Implementation of the Synthetic Turbulence Generation (STG) algorithm
+/// presented in Shur et al. 2014
+//
+/// SetupSTG_Rand reads in the input files and fills in STGShur14Context. Then
+/// STGShur14_CalcQF is run over quadrature points. Before the program exits,
+/// TearDownSTG is run to free the memory of the allocated arrays.
+
+#ifndef stg_shur14_h
+#define stg_shur14_h
+
+#include <math.h>
+#include <ceed.h>
+#include <stdlib.h>
+
+#ifndef stg_shur14_struct
+#define stg_shur14_struct
+typedef struct STGShur14Context_ *STGShur14Context;
+struct STGShur14Context_ {
+  CeedInt nmodes;              // !< Number of wavemodes
+  CeedScalar *sigma, *d, *phi; // !< Random number set, [nmodes,3], [nmodes,3], [nmodes]
+  CeedScalar *kappa;           // !< Wavemode frequencies, [nmodes]
+  CeedScalar alpha;            // !< Geometric growth rate of kappa
+
+  CeedInt nprof;       // !< Number of profile points in STGInflow.dat
+  CeedScalar *prof_dw; // !< Distance to wall for Inflow Profie, [nprof]
+  CeedScalar *ubar;    // !< Mean velocity, [nprof, 3]
+  CeedScalar *cij;     // !< Cholesky decomposition [nprof, 6]
+  CeedScalar *eps;     // !< Turbulent Disspation [nprof, 6]
+  CeedScalar *lt;      // !< Tubulent Length Scale [nprof, 6]
```

Comment:
What you're doing here will work fine on a CPU, but it will need some changes for a GPU. Not a big deal to do then, but if you want to do it now, one approach is
```suggestion
  CeedInt nmodes;              // !< Number of wavemodes
  CeedScalar alpha;            // !< Geometric growth rate of kappa
  CeedInt nprof;       // !< Number of profile points in STGInflow.dat
  struct {
    CeedInt sigma, d, phi;
    CeedInt prof_dw, ubar, cij, eps, lt;
  } offsets;
  CeedScalar data[];
```

and then access as
```c
const CeedScalar *sigma = &ctx->data[ctx->offsets.sigma];
```
This is called a [flexible array member](https://en.wikipedia.org/wiki/Flexible_array_member). Note that it's only strictly a C feature, but in practice, g++ and clang++ accept it at `-Wall -Wextra`, and only warn at `-Wpedantic`.

Reply:
Ok, cool. Just so I understand correctly, if we assume that the data for `sigma`, `d`, and `phi` follow each other (which are sizes `[nmodes, 3]`, `[nmodes,3]`, `[nmodes]`, repsectively), then the offsets would be:

```c
ctx->offsets.sigma = 0
ctx->offsets.d = ctx->nmodes * 3
ctx->offsets.phi = ctx->offsets.d + ctx->nmodes*3
ctx->offsets.prof_dw = ctx->offsets.phi + ctx->nmodes // Or whatever data comes after phi
```
?

Reply:
Yes to both. Sorry it's a bit clumsy -- we could give it some interface sugar, but let's not get distracted by that just now.

Reply:
No worries. I just wasn't familiar with that construct. I believe I've resolved this though with the most recent commits.